### PR TITLE
client: consider inode's last read/write when calculating wanted caps

### PR DESCRIPTION
**Pull request by ukernel**
Add last_rd and last_wr to struct Inode. These fields are used to track
the last time the client acquired read/write caps for the inode.

If there is no read/write on an inode for 'caps_wanted_delay_max'
seconds, __ceph_caps_file_wanted() does not request caps for read/write
even there are open files.

For dir operations. __ceph_caps_file_wanted() calculates dir's wanted
caps according to last dir read/modification. If there is recent dir
read dir inode wants CEPH_CAP_ANY_SHARED caps. If there is recent dir
modification, also wants CEPH_CAP_FILE_EXCL.

Readdir is a special case. Dir inode wants CEPH_CAP_FILE_EXCL after
readdir, as with that, modifications do not need to release
CEPH_CAP_FILE_SHARED or invalidate all dentry leases issued by readdir.

Fixes: https://tracker.ceph.com/issues/43748
Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ukernel** [comment]
http://pulpito.ceph.com/zyan-2020-03-19_12:10:02-fs-wip-open-file-timeout-distro-basic-smithi/4868009

**Comment by ukernel** [comment]
jenkins retest this please

**Comment by batrick** [comment]
retest this please

File: qa/tasks/cephfs/test_misc.py [hunk 1]
```diff
@@ -187,6 +190,61 @@ def test_dump_inode_hexademical(self):
         info = self.fs.mds_asok(['dump', 'inode', hex(ino)])
         assert info['path'] == "/foo"
 
+    def test_open_file_timeout(self):
```
Comment (batrick) [inline qa/tasks/cephfs/test_misc.py:193]: Add test description.

File: qa/tasks/cephfs/test_misc.py [hunk 2]
```diff
@@ -187,6 +190,61 @@ def test_dump_inode_hexademical(self):
...
+        if not isinstance(self.mount_a, FuseMount):
+            self.skipTest("Require FUSE client")
+
+        self.mount_a.umount_wait()
+
+        timeout = 30
+        self.set_conf('client', 'client caps wanted delay max', timeout)
```
Comment (batrick) [inline qa/tasks/cephfs/test_misc.py:200]: `self.config_set`

File: src/common/options.cc [hunk 3]
```diff
@@ -8249,10 +8249,14 @@ std::vector<Option> get_mds_client_options() {
     .set_default(30)
     .set_description(""),
 
-    Option("client_caps_release_delay", Option::TYPE_INT, Option::LEVEL_DEV)
+    Option("client_caps_wanted_delay_min", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
     .set_default(5)
     .set_description(""),
 
+    Option("client_caps_wanted_delay_max", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    .set_default(60)
```
Comment (batrick) [inline src/common/options.cc:8257]:
```suggestion
    .set_default(60)
    .set_min(??)
```

File: src/common/options.cc [hunk 4]
```diff
@@ -8249,10 +8249,14 @@ std::vector<Option> get_mds_client_options() {
...
+    .set_description(""),
```
Comment (batrick) [inline src/common/options.cc:8258]: Please add a description. (Only LEVEL_DEV is optional.)

File: src/common/options.cc [hunk 5]
```diff
@@ -8249,10 +8249,14 @@ std::vector<Option> get_mds_client_options() {
     .set_default(30)
     .set_description(""),
 
-    Option("client_caps_release_delay", Option::TYPE_INT, Option::LEVEL_DEV)
+    Option("client_caps_wanted_delay_min", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
     .set_default(5)
     .set_description(""),
```
Comment (batrick) [inline src/common/options.cc:8254]: ditto

**Comment by batrick** [comment]
jenkins test dashboard backend

**Comment by ukernel** [comment]
close this in favor of https://github.com/ceph/ceph/pull/32576

**Pull request closed without merging**